### ISSUE TITLE
symlinked gitignore file fails with "fatal: cannot use <path> as an exclude file"

### ISSUE THREAD

**Issue by 0xabu** [git-for-windows/git#1392]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options
git version 2.15.1.windows.2
built from commit: 5d5baf91824ec7750b103c8b7c4827ffac202feb
sizeof-long: 4
machine: x86_64
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver
Microsoft Windows [Version 10.0.16299.98]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
> type "c:\Program Files\Git\etc\install-options.txt"
Editor Option: VIM
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: OpenSSL
CRLF Option: CRLFAlways
Bash Terminal Option: MinTTY
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Disabled
```

However, I have also tried manually changing the symlink setting:
```
> type c:\ProgramData\Git\config
[core]
        symlinks = true
        autocrlf = true
        fscache = true
(...)
```

```
> git config --global --get core.symlinks
true
```

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

This repros from both CMD and Cygwin ZSH

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

I try to keep all my configuration files in OneDrive, and use symlinks to point to them. I have both ~/.gitconfig and ~/.gitignore in OneDrive. Cygwin and WSL git are both happy with this arrangement, but Git for Windows ignores and refuses to use a symlinked .gitignore file (even if it's still happy with a symlink at ~/.gitconfig).

Here's a self-contained repro, starting from `C:\Users\<me>`:
```
> echo "somepattern" > .gitignore
> git config --global core.excludesfile ~/.gitignore
> type .gitconfig
[core]
        excludesfile = ~/.gitignore
> mkdir tmp
> cd tmp
> git init
Initialized empty Git repository in c:/Users/<me>/tmp/.git/
> git status
On branch master

No commits yet

nothing to commit (create/copy files and use "git add" to track)
> cd ..
> mv .gitignore thegitignore
> mklink .gitignore thegitignore
symbolic link created for .gitignore <<===>> thegitignore
> cd tmp
> git status
fatal: cannot use C:/Users/<me>/.gitignore as an exclude file
```
 - What did you expect to occur after running these commands?

Git reads and uses my custom .gitignore file, just as it did in the first instance of "git status" when it wasn't a symlink.

 - What actually happened instead?

Git prints the warning:
```
fatal: cannot use C:/Users/<me>/.gitignore as an exclude file
```
and ignores the contents of that file.

A syscall trace shows that git is actually following the symlink and reading the contents of the linked file before ignoring it, so it's unclear what exactly is wrong with having a symlink there:

![image](https://user-images.githubusercontent.com/13037682/33681963-ec093b5e-da7a-11e7-9b4e-caa2f59924a6.png)

**Comment by ralish** [comment]
Seeing the exact same behaviour. Important to note that this is a regression. I've been using a symlinked `.gitignore` file for years without issue until I updated to Git 2.15.1. I'm pretty sure I was on 2.13 previously.

**Comment by shiftkey** [comment]
We've had a report of this in GitHub Desktop - https://github.com/desktop/desktop/issues/3617 - I've not had a chance to dig into where this was introduced, but b2353379bba414e6c00dde913497cc9c827366f2 is an interesting recent change because this error is due to calling `add_excludes` and encountering some error path:

```c
if (add_excludes(fname, "", 0, el, NULL, sha1_stat) < 0)
    die("cannot use %s as an exclude file", fname);
```

cc @jeffhostetler for any other ideas

**Comment by jeffhostetler** [comment]
I'm looking at it now.

**Comment by jeffhostetler** [comment]
I found the problem.  The subtleties of lstat() vs stat().  I posted a fix here:

    https://github.com/git-for-windows/git/pull/1407

**Comment by dscho** [comment]
Fixed via #1407. @0xabu, once the snapshot is available at https://wingit.blob.core.windows.net/files/index.html (should not take more than an hour), would you mind verifying that it is fixed?

**Comment by dscho** [comment]
> @0xabu, once the snapshot is available at https://wingit.blob.core.windows.net/files/index.html (should not take more than an hour), would you mind verifying that it is fixed?

@0xabu There are now *multiple* snapshots with the fix, so can you please test that this ticket can be closed? Thank you.

**Comment by 0xabu** [comment]
@dscho I can confirm this is fixed in Git-prerelease-2.15.1.windows.2.27.g512e5d63ec-64-bit -- thanks all!

**Comment by dscho** [comment]
> I can confirm this is fixed in Git-prerelease-2.15.1.windows.2.27.g512e5d63ec-64-bit

Thanks for the confirmation!
